### PR TITLE
Move identity configuration services.xml -> deployment.xml

### PR DESCRIPTION
**Pull request by tokle**
fyi: @bjorncs, @freva

I touched the config-model-api, but should be only additions...

### CODE REVIEW

File: config-model-api/src/main/java/com/yahoo/config/application/api/DeploymentSpec.java [hunk 1]
```diff
@@ -81,7 +89,30 @@ private void validateZones(List<Step> steps) {
...
+    private void validateAthenz() {
+        // If athenz domain is not set, athenz service cannot be set on any level
+        if (! athenzDomain.isPresent()) {
+            for (DeclaredZone zone : zones()) {
+                if(zone.athenzService().isPresent()) {
+                    throw new IllegalArgumentException("Athenz service configured for zone: " + zone + ", but Athenz domain is not configured");
+                }
+            }
+        // if athenz domain is configured, athenz service must be set implicitly or directly on all zones.
+        } else if(! athenzService.isPresent()) {
```
Comment (gjoranv) [inline config-model-api/src/main/java/com/yahoo/config/application/api/DeploymentSpec.java]: athenzService or athenzDomain??

File: config-model/src/test/java/com/yahoo/vespa/model/container/xml/IdentityBuilderTest.java [hunk 2]
```diff
@@ -19,16 +23,50 @@
     @Test
     public void identity_config_produced() throws IOException, SAXException {
```
Comment (gjoranv) [inline config-model/src/test/java/com/yahoo/vespa/model/container/xml/IdentityBuilderTest.java]: Test name should constitute a requirement, e.g. (Require that) identity_config_is_produced_when_deployment_xml_specifies_athenz

File: config-model/src/test/java/com/yahoo/vespa/model/container/xml/IdentityBuilderTest.java [hunk 3]
```diff
@@ -19,16 +23,50 @@
...
+        root = new MockRoot("root", applicationPackage);
+        createModel(root, DeployState.createTestState(applicationPackage), clusterElem);
+
         IdentityConfig identityConfig = root.getConfig(IdentityConfig.class, "default/component/" + Identity.CLASS);
         assertEquals("domain", identityConfig.domain());
         assertEquals("service", identityConfig.service());
     }
+
+    @Test
+    public void athenz_service_override() throws IOException, SAXException {
```
Comment (gjoranv) [inline config-model/src/test/java/com/yahoo/vespa/model/container/xml/IdentityBuilderTest.java]: Again, test should have a name that is a requirement, e.g. athenz_service_can_be_overridded_for_...

File: config-model/src/main/java/com/yahoo/vespa/model/container/ContainerCluster.java [hunk 4]
```diff
@@ -304,6 +307,38 @@ public void addJaxProviders() {
         addSimpleComponent(XPathFactoryProvider.class);
     }
 
+    public void addIdentity() {
```
Comment (gjoranv) [inline config-model/src/main/java/com/yahoo/vespa/model/container/ContainerCluster.java]: Please refactor, this method is too long. Most of this code seems irrelevant to ContainerCluster. If I read correctly, it just sets two properties on the cluster, which could perhaps be set from the builder. I'm not sure, but let's look at this again after this method has been refactored to become more legible.

File: config-model/src/main/java/com/yahoo/vespa/model/container/ContainerCluster.java [hunk 5]
```diff
@@ -304,6 +307,38 @@ public void addJaxProviders() {
...
+        if (getDeploymentSpec().isPresent()) {
+            DeploymentSpec deploymentSpec = getDeploymentSpec().get();
```
Comment (gjoranv) [inline config-model/src/main/java/com/yahoo/vespa/model/container/ContainerCluster.java]: These two lines can be replaced by 
```
getDeploymentSpec().map(deploymentSpec -> {
   ...
})
```

File: config-model/src/main/java/com/yahoo/vespa/model/container/ContainerCluster.java [hunk 6]
```diff
@@ -304,6 +307,38 @@ public void addJaxProviders() {
...
+            if (deploymentSpec.athenzDomain().isPresent()) {
+                String domain = deploymentSpec.athenzDomain().get();
```
Comment (gjoranv) [inline config-model/src/main/java/com/yahoo/vespa/model/container/ContainerCluster.java]: Same here

File: config-model/src/main/java/com/yahoo/vespa/model/container/ContainerCluster.java [hunk 7]
```diff
@@ -304,6 +307,38 @@ public void addJaxProviders() {
...
+                // Service can be configured globally, or per zone
```
Comment (gjoranv) [inline config-model/src/main/java/com/yahoo/vespa/model/container/ContainerCluster.java]: This could perhaps be a static utility method in Identity?

File: config-model/src/main/java/com/yahoo/vespa/model/container/ContainerCluster.java [hunk 8]
```diff
@@ -304,6 +307,38 @@ public void addJaxProviders() {
...
+                String service = deploymentSpec.zones().stream()
+                        .filter(zone -> zone.deploysTo(getZone().environment(), Optional.of(getZone().region())))
+                        .findAny()
+                        .flatMap(DeploymentSpec.DeclaredZone::athenzService)
+                        .orElse(deploymentSpec.athenzService().orElseThrow(() -> new RuntimeException("Missing Athenz service configuration")));
+
+                // Set lbaddress, or use first hostname if not specified.
```
Comment (gjoranv) [inline config-model/src/main/java/com/yahoo/vespa/model/container/ContainerCluster.java]: This also seems like it could be a utility in Identity.

File: config-model/src/main/java/com/yahoo/vespa/model/container/ContainerCluster.java [hunk 9]
```diff
@@ -304,6 +307,38 @@ public void addJaxProviders() {
...
+                HostName lbName = Optional.ofNullable(getRoot().getDeployState().getProperties().loadBalancerName())
+                        .orElseGet(
+                                () -> HostName.from(getRoot().getDeployState().getProperties().configServerSpecs().stream()
+                                                            .findFirst()
+                                                            .map(ConfigServerSpec::getHostName)
+                                                            .orElse("unknown") // Currently unable to test this, hence the unknown
+                                ));
+
+                Identity identity = new Identity(domain.trim(), service.trim(), lbName);
```
Comment (gjoranv) [inline config-model/src/main/java/com/yahoo/vespa/model/container/ContainerCluster.java]: If 'Identity' is a representation of the 'IdentityProvider' component, I think we should rename the config model class as well.

**Review by gjoranv — CHANGES_REQUESTED** [review]

File: config-model/src/main/java/com/yahoo/vespa/model/container/ContainerCluster.java (hunk 5 above)
Comment (tokle) [reply config-model/src/main/java/com/yahoo/vespa/model/container/ContainerCluster.java]: I know, just didn't feel right at once, but OK

File: config-model-api/src/main/java/com/yahoo/config/application/api/DeploymentSpec.java [hunk 10]
```diff
@@ -81,7 +89,30 @@ private void validateZones(List<Step> steps) {
...
+     */
+    private void validateAthenz() {
+        // If athenz domain is not set, athenz service cannot be set on any level
+        if (! athenzDomain.isPresent()) {
+            for (DeclaredZone zone : zones()) {
+                if(zone.athenzService().isPresent()) {
+                    throw new IllegalArgumentException("Athenz service configured for zone: " + zone + ", but Athenz domain is not configured");
+                }
+            }
+        // if athenz domain is configured, athenz service must be set implicitly or directly on all zones.
```
Comment (gjoranv) [inline config-model-api/src/main/java/com/yahoo/config/application/api/DeploymentSpec.java]: // Athenz domain is set, so ...

**Review by gjoranv — APPROVED** [review]

File: config-provisioning/src/main/java/com/yahoo/config/provision/AthenzService.java [hunk 11]
```diff
@@ -0,0 +1,20 @@
+// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+package com.yahoo.config.provision;
+
+/**
+ * @author mortent
+ */
+public class AthenzService {
```
Comment (gjoranv) [inline config-provisioning/src/main/java/com/yahoo/config/provision/AthenzService.java]: Please add equals + hashCode + toString.

File: config-provisioning/src/main/java/com/yahoo/config/provision/AthenzDomain.java [hunk 12]
```diff
@@ -0,0 +1,20 @@
+// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+package com.yahoo.config.provision;
+
+/**
+ * @author mortent
+ */
+public class AthenzDomain {
```
Comment (gjoranv) [inline config-provisioning/src/main/java/com/yahoo/config/provision/AthenzDomain.java]: Please add equals + hashCode + toString.

File: config-model-api/src/test/java/com/yahoo/config/application/api/DeploymentSpecTest.java [hunk 13]
```diff
@@ -368,7 +368,7 @@ public void deploymentSpecWithChangeBlocker() {
     }
 
     @Test
-    public void deploymentSpecWithAthenzIdentity() {
+    public void athenz_config_is_read_from_deployment() {
```
Comment (gjoranv) [inline config-model-api/src/test/java/com/yahoo/config/application/api/DeploymentSpecTest.java]: great 🏆

**Review by gjoranv — CHANGES_REQUESTED** [review]

**Review by gjoranv — APPROVED** [review]

**Comment by tokle** [comment]
@hmusum: please have a look at the last commit

**Comment by hmusum** [comment]
Last commit looks good to me

**Comment by gjoranv** [comment]
🥇
